client/modules/wj-util/src/index.ts

```diff
@@ -0,0 +1,159 @@
+  let h = 0
+  for (let i = 0; i < s.length; i++) {
+    h = (Math.imul(31, h) + s.charCodeAt(i)) | 0
+  }
+  return h
+}
+
+export interface SearchOpts {
+  /** Starting minimum index for the search. */
+  min?: number
+  /** Starting maximum index for the search. */
+  max?: number
+  /** If true, the search will return the closest index to the
+   *  desired value on failure. */
+  precise?: boolean
+}
+
+/** Performs a binary search through an array.
+ *
+ *  The comparator function should return -1 if undershooting the desired value,
+ *  +1 if overshooting, and 0 if the value was found.
+ *
+ *  The comparator can also short-circuit the search by returning true or false.
+ *  Returning true is like returning a 0 (target found), but
+ *  returning false induces a null return. */
+export function search<T, TR>(
+  haystack: T[],
+  target: TR,
+  comparator: (element: T, target: TR) => number | boolean,
+  { min = 0, max = haystack.length - 1, precise = true }: SearchOpts = {}
+) {
+  let index = -1
+  while (min <= max) {
+    index = min + ((max - min) >>> 1)
+    const cmp = comparator(haystack[index], target)
+    if (cmp === true || cmp === 0) return { element: haystack[index], index }
+    if (cmp === false) return null
+    if (cmp < 0) min = index + 1
+    else if (cmp > 0) max = index - 1
+  }
+
+  if (!precise) return { element: null, index }
+
+  return null
+}
+
+/** Checks if an array or object is empty. Will return true for non-objects. */
+export function isEmpty(obj: any) {
+  if (!obj) return true
+  if (obj instanceof Array) return obj.length === 0
+  if (obj.constructor === Object) return Object.keys(obj).length === 0
+  return true
+}
+
+/** Creates a type that is the type of `T` if it had a known property `K`. */
+type Has<K extends string, T> = T extends { [P in K]?: infer R }
+  ? Omit<T, K> & Record<K, R>
+  : never
+
+/** Returns if an object `T` has a key `K`, and only returns true if
+ *  the value of that key isn't undefined. */
+export function has<K extends string, T>(
+  key: K,
+  obj: T
+): obj is T extends Record<any, any> ? Has<K, T> : never {
+  if (!isAnyObject(obj)) return false
+  return key in obj && (obj as any)[key] !== undefined
+}
+
+/** Removes all properties assigned to `undefined` in an object. */
+export function removeUndefined<T>(obj: T) {
+  // this wacky approach is faster as it avoids an iterator
+  const keys = Object.keys(obj) as (keyof T)[]
+  for (let i = 0; i < keys.length; i++) {
+    if (obj[keys[i]] === undefined) delete obj[keys[i]]
+  }
+  return obj as { [K in keyof T]: Exclude<T[K], undefined> }
+}
+
+/** Takes a string and escapes any `RegExp` sensitive characters. */
+export function escapeRegExp(str: string) {
+  return str.replace(/[.*+?^${}()|\[\]\\]/g, "\\$&")
+}
+
+/** Checks if a string has any of the provided sigils.
+ *  (e.g. `hasSigil('!string', '!') -> true`) */
+export function hasSigil<T extends string = string>(
+  str: unknown,
```

Comment:
That code checks if `str` is a string or not. It's so that you can pass in objects willy-nilly and not worry about their type. Tarnation uses this a lot when demangling the grammar definition, it's a really useful feature.